doc/Hammer-Use/Hammer-Config.rst

```diff
@@ -211,6 +211,20 @@ Common meta actions
 
   Result: ``get_setting("foo.bar.baz")`` returns ``12345`` and ``get_setting("foo.bar.baz")`` returns ``32123``
 
+Type Checking
+-------------
+
+Any existing configuration file can and should be accompanied with a corresponding configuration types file.
+This allows for static type checking of any key when calling ``get_setting``.
+The file should contain the same keys as the corresponding configuration file, but can contain the following as values:
+
+- primitive types (``int``, ``str``, etc.)
+- collection types (``list``)
+- collections of key-value pairs (``list[dict[str, str]]``) to ensure compatibility with custom constraints
```

Comment:
I think that this needs to be a bit clearer - "to ensure compatibility" doesn't seem like enough. I think we'd need to say like "these dicts are to be translated into custom constraints (PlacementConstraint, PinAssignment, etc.) but are not themselves checked for types within the key-value pairs. See `src/hammer-vlsi/hammer_vlsi/constraints.py` for the supported custom constraints to ensure that you are specifying them correctly.

Reply:
Custom constraints are either `list[dict[str, str]]` or `list[dict[str, list]]`. The first accounts for most custom constraints, and the second accounts specifically for manual placement constraints and other hierarchical constraints. Either way, it's difficult to test comprehensively since the defaults are usually empty.